File: strategy/generator/signal/signalline/di_signalline.py
from dataclasses import dataclass

from core.models.parameter import (
    Parameter,
    RandomParameter,
    StaticParameter,
)
from core.models.smooth import Smooth
from strategy.generator.signal.base import Signal, SignalType


@dataclass(frozen=True)
class DiSignalLineSignal(Signal):
    type: SignalType = SignalType.DiSignalLine
    period: Parameter = RandomParameter(10.0, 15.0, 1.0)
    smooth_type: Parameter = StaticParameter(Smooth.WMA)
    signal_period: Parameter = RandomParameter(4.0, 8.0, 1.0)

File: ta_lib/strategies/signal/src/signalline/di_signalline.rs
use base::prelude::*;
use core::prelude::*;
use momentum::di;

pub struct DiSignalLineSignal {
    smooth_type: Smooth,
    period: usize,
    signal_period: usize,
}

impl DiSignalLineSignal {
    pub fn new(smooth_type: Smooth, period: f32, signal_period: f32) -> Self {
        Self {
            smooth_type,
            period: period as usize,
            signal_period: signal_period as usize,
        }
    }
}

impl Signal for DiSignalLineSignal {
    fn lookback(&self) -> usize {
        std::cmp::max(self.period, self.signal_period)
    }

    fn generate(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let di = di(&data.close(), self.smooth_type, self.period);
        let signal_line = di.smooth(self.smooth_type, self.signal_period);

        (di.cross_over(&signal_line), di.cross_under(&signal_line))
    }
}
